### PR TITLE
Timespinner: many new stuffs

### PR DESCRIPTION
**Pull request by Jarno458**
## What is this fixing or adding?
(Only listing the AP changes here, not client changes)
New Flags:
  * [Dad Percent]: "The win condition is beating the boss of Emperor's Tower"
  * [Rising Tides]: "Random area's are flooded"
    * Possible areas that can be flooded:
      * Maw caves + boss spot
      *  Xarion boss spot
      * Lake Desolation chicken jump
      *  Castle Courtyard
      *  Castle Basement
      * Castle Moat
      *  Ancient Pyramid Shaft + regrets chest
      * Ancient Pyramid boss spot
      * Dry Lake Serene
  * [Unchained Keys]: "Start with Twin Pyramid Key, which does not give free warp; warp items for Past, Present, (and ??? with Enter Sandman) can be found"
  * [Trapped Chests]: "Items can be traps. Toggles available traps in the 'Traps' settings"
  * [LevelCap]: "Sets the max level Lunais can achieve."
  * [ExtraEarringsXP]: "Adds additional XP granted by Galaxy Earrings."
     
Bug Fixes:
  * Fixed game from crashing when playing on nightmare lvl1 without first clearing the game
  * Fixed connecting with the wrong save to archipelago multiworld

New Yaml options:
```yaml
  LevelCap: 80 #Range 1 - 99 (default: 99)
  ExtraEarringsXP: 5 #Range 0 - 24 (default: 0)
  DadPercent: random #Toggle (default: off)
  RisingTides: true #Toggle (default: off)
  RisingTidesOverrides: # Default weights "Dry": 67, "Flooded": 33
    Xarion: # Regions can be: CastleBasement, Xarion, Maw, AncientPyramidShaft, Sandman, CastleMoat, CastleCourtyard, LakeDesolation, LakeSerene
      Dry: 50
      Flooded: 450
  UnchainedKeys: random #Toggle (default: off)
  TrapChance: 100 # Range 0 - 100 (default: 10%)
  Traps: # Default: all
    - 'Bee Trap' # Can bee: 'Meteor Sparrow Trap', 'Poison Trap', 'Chaos Trap', 'Neurotoxin Trap', 'Bee Trap'
```

## How was this tested?
Using AP world available in the timespinner channel in discord channel for the last two weeks
Also many generation tests, to find generation issues
Tracker changes where tested through a local webhost only
(the client side of the changes where tested and developed over past two months as a beta)

### CODE REVIEW

**Review by alwaysintreble — COMMENTED** [review]
Looks good other than a random change I don't understand. Would be nice if you added some tests for the new logic stuff -hint hint-

File: setup.py [hunk 1]
```diff
@@ -45,6 +45,7 @@
     "Rogue Legacy",
     "Donkey Kong Country 3",
     "Super Mario World",
+    "Timespinner"
```
Comment (Berserker66) [inline setup.py:48]:
```suggestion
    "Timespinner",
```

File: worlds/timespinner/__init__.py [hunk 2]
```diff
@@ -102,19 +108,78 @@ def generate_basic(self):
...
+                spoiler_handle.write('Mysterious Warp Beacon unlock:   %s\n' % 
+                    (self.precalculated_weights.time_key_unlock))
+        else:
+            spoiler_handle.write('Twin Pyramid Keys unlock:        %s\n' % 
+                (self.precalculated_weights.pyramid_keys_unlock))
+       
+        if is_option_enabled(self.multiworld, self.player, "RisingTides"):
+            flooded_areas: List[str] = []
+
+            if(self.precalculated_weights.flood_basement):
```
Comment (black-sliver) [inline worlds/timespinner/__init__.py:155]: Please consider removing the `()` from the ifs here

File: worlds/timespinner/__init__.py [hunk 3]
```diff
@@ -71,28 +68,37 @@ def generate_early(self):
...
 
     def create_item(self, name: str) -> Item:
         return create_item_with_correct_settings(self.multiworld, self.player, name)
 
     def get_filler_item_name(self) -> str:
-        return self.multiworld.random.choice(filler_items)
+        trap_chance: int = get_option_value(self.multiworld, self.player, "TrapChance")
+        enabled_traps: List[str] = get_option_value(self.multiworld, self.player, "Traps")
+
+        if (self.multiworld.random.random() < (trap_chance / 100) and enabled_traps):
```
Comment (black-sliver) [inline worlds/timespinner/__init__.py:81]:
```suggestion
        if self.multiworld.random.random() < (trap_chance / 100) and enabled_traps:
```

File: worlds/timespinner/__init__.py [hunk 4]
```diff
@@ -102,19 +108,78 @@ def generate_basic(self):
...
+        slot_data["LakeDesolation"] = self.precalculated_weights.flood_lake_desolation
+        slot_data["DryLakeSerene"] = self.precalculated_weights.dry_lake_serene
 
         return slot_data
 
     def write_spoiler_header(self, spoiler_handle: TextIO):
-        spoiler_handle.write('Twin Pyramid Keys unlock:        %s\n' % (self.pyramid_keys_unlock))
+        if is_option_enabled(self.multiworld, self.player, "UnchainedKeys"):
+            spoiler_handle.write('Modern Warp Beacon unlock:       %s\n' % 
+                (self.precalculated_weights.present_key_unlock))
```
Comment (black-sliver) [inline worlds/timespinner/__init__.py:141]: Also the way you pass in the strings looks wrong:
Having `(something)` looks like it should be `(something,)` (i.e. a tuple). If you explicitly just want `something` (i.e. a str), then remove the `()` - sounds like a small thing, but having to figure out what a pair of `()` does makes reviewing a bit harder.

**Review by black-sliver — CHANGES_REQUESTED** [review]
Tracker requires https://github.com/ArchipelagoMW/Archipelago/pull/1453 but seems to work fine with it.

Found a couple of typos and some inconsistencies. I don't really care where you put your braces and brackets and white spaces, but it should be consistent.

Also please make sure the changes in WebHostLib match the corresponding sections from style.md. (You added a lot of spaces in front of `{` in the css)

(Looks like  #1034 will pass after merging this)

**Comment by Jarno458** [comment]
> Tracker requires #1453 but seems to work fine with it.
WDYM requires? this was tested and developed without any addition branches, just master + my changes?

**Comment by black-sliver** [comment]
> > Tracker requires #1453 but seems to work fine with it.
> > @black-sliver WDYM requires? this was tested and developed without any addition branches, just master + my changes?

Tracker is currently broken on main, not your fault, so when you merged main, it broke on your branch as well. Just wanted to point that out if anyone else wants to test.

**Review by black-sliver — APPROVED** [review]

**Pull request merged**